annotation_converters/objectron_2_coco.py

```diff
@@ -0,0 +1,198 @@
+                vid_path = osp.join(data_root, 'videos' + osp.sep + line.strip() + osp.sep + 'video.MOV')
+                videos_info.append((vid_path, ann))
+
+    avg_vid_len /= len(videos_info)
+    return videos_info, avg_vid_len
+
+
+def np_encoder(object):
+    if isinstance(object, np.generic):
+        return object.item()
+
+
+def decode_keypoints(keypoints, keypoint_size_list, size):
+    keypoints = np.split(keypoints, np.array(np.cumsum(keypoint_size_list)))
+    keypoints = [points.reshape(-1, 3) for points in keypoints]
+    unwrap_mat = np.asarray([size[0], size[1], 1.], np.float32)
+    keypoints = [
+        np.multiply(keypoint, unwrap_mat).astype(int)[:, :-1]
+            for keypoint in keypoints
+    ][:len(keypoint_size_list)]
+    return keypoints
+
+
+def get_bboxes_from_keypoints(keypoints, num_objects, size):
+    w, h = size
+    bboxes = []
+    num_valid = 0
+
+    for i in range(num_objects):
+        min_x = np.min(keypoints[i][:,0])
+        min_y = np.min(keypoints[i][:,1])
+        max_x = np.max(keypoints[i][:,0])
+        max_y = np.max(keypoints[i][:,1])
+        bbox = [min_x, min_y, max_x - min_x, max_y - min_y]
+        if min_x < 0 or min_y < 0 or max_x >= w or max_y >= h and bbox[2]*bbox[3] == 0:
+            bboxes.append(None)
+        else:
+            bboxes.append(bbox)
+            num_valid += 1
+
+    if num_valid > 0:
+        return bboxes
+
+    return None
+
+
+def save_2_coco(output_root, subset_name, data_info, obj_classes, fps_divisor, res_divisor, dump_images=False):
+    json_name = f'objectron_{subset_name}.json'
+    ann_folder = osp.join(output_root, 'annotations')
+    img_folder = osp.join(output_root, 'images')
+    if not osp.isdir(ann_folder):
+        os.mkdir(ann_folder)
+    if not osp.isdir(img_folder):
+        os.mkdir(img_folder)
+
+    img_id = 0
+    ann_id = 0
+
+    stat = {'Total frames' : 0, 'Avg box size': [0, 0], 'Total boxes': 0}
+    ann_dict = {}
+    categories = [{"id": i + 1, "name": cl} for i, cl in enumerate(obj_classes)]
+    class_2_id = {cl : i + 1 for i, cl in enumerate(obj_classes)}
+    images_info = []
+    annotations = []
+
+    for item in tqdm(data_info):
+        vid_path, annotation = item
+        # assert get_video_frames_number(vid_path) == len(annotation)
+        req_frames = []
+        for frame_idx in range(len(annotation)):
+            if frame_idx % fps_divisor == 0:
+                req_frames.append(frame_idx)
+        frames = grab_frames(vid_path, req_frames)
+
+        for frame_idx, frame_ann in enumerate(annotation):
+            if frame_idx not in frames:
+                continue
+            if frames[frame_idx] is None:
+                print('Warning: missing frame in ' + vid_path)
+                continue
+            #object_keypoints_2d, object_categories, keypoint_size_list, annotation_types
+
+            h, w = frames[frame_idx].shape[0] // res_divisor, frames[frame_idx].shape[1] // res_divisor
```

Comment:
What if res_divisor is not 1? As I understand, then, the shape of the image would be changed and decoding of bbox coordinates would be with the new resolution, but they were encoded with the old one.

Reply:
The formula of decoding in `decode_keypoints` is linear, so I guess I can decode with any resolution. Also I visually checked annotation with `res_divisor=2` and currently training a detector on it.

Reply:
Yes, seems so, I agree since it's linear